---
---
<!doctype html>
<html>
  {% include header.html %}
  <body>
    <div class="wrapper">
      {% include menu.html %}
      <section>
      <h2>Best practices for using lobSTR with BWA-MEM alignments</h2>
       <ul>
      <li><a href="#overview">Overview</a></li>
      <li><a href="#issues">Caveats when using external aligners</a></li>
      <li><a href="#options">Recommended allelotype options</a></li>
      </ul>
  <p></p>
    <a name="overview"></a>
    <h2>Overview</h2>
    <p>
           The original lobSTR pipeline involves running two steps:
      <ul>
      	<li><code>lobSTR</code>: Detects repetitive reads and aligns them to STRs in the genome, outputting a BAM file</li>
      	<li><code>allelotype</code>: Takes a BAM file generated by <code>lobSTR</code> and produces genotype calls at each STR</li>
      </ul>
      When lobSTR was developed, existing aligners like <code>bwa aln</code> and <code>bowtie</code> were very insentitive to large insertions or deletions from the reference genome. Therefore, these aligners missed a large fraction of polymorphism at STRs and we needed to create an aligner specific to STR regions. However, the more recently developed <a href="http://bio-bwa.sourceforge.net/">BWA-MEM</a>	has proven to be quite sensitive to large indels. It is now possible to use BAMs generated by <code>BWA-MEM</code> as input to <code>allelotype</code>.

      This has a number of advantages:
      <ul>
      	<li>Large cohorts have already been aligned by <code>BWA-MEM</code>, saving the computational cost of running <code>lobSTR</code>.</li>
      	<li><code>BWA-MEM</code> alignment is generally more sensitive, resulting in higher coverage for making STR calls.</li>
      </ul>

      A number of caveats of using external aligners are explained below. We also provide recommended options to <code>allelotype</code> when using <code>BWA-MEM</code>.
	</p>
	<a name="issues"></a>
    <h2>Caveats when using external aligners</h2>
    <p>
    Different aligners behave differently in terms of their alignment algorithms, output formats, and scoring conventions. The following points are caveats that we found using other aligners:
    <ul>
    <li><em>Soft-clipped reads</em>: If the ends of a read does not align well but the rest of the read maps to a location unambiguously, <code>BWA-MEM</code> will sometimes soft clip the end of the read. At STR regions, especially those in which the region flanking the STR also happens to be a bit messy, this can result in a read being called as the reference allele when in reality there is a length variation, but the variable region just happened to get clipped. To remedy this, we suggest either re-doing local realignment by adding the <code>--realign</code> option, or simply filtering these reads using <code>--filter-clipped</code>.
    <li><em>Multiple mappers</em>: By default, <code>lobSTR</code> discards reads that map to multiple locations. Many aligners, including <code>BWA-MEM</code> mark reads that do not have a unique best alignment with a map quality score of 0. To filter these reads from allelotyping, use the flag <code>--filter-map0</code>.</li>
    <li><em>Messy read ends</em>: to use a read for allelotyping, we require a read to fully span the STR of interest. <code>BWA-MEM</code> is often much more sensitive, as it aligns the whole read at once rather than relying on finding alignments for each flanking region separately. As a result, a lot of reads with a flanking region too messy to be mapped by <code>lobSTR</code> will end up being mapped by <code>BWA-MEM</code>. This is mostly a benefit, as it allows making calls based on higher coverage. However, reads that barely span an STR and have messy flanking regions are much more likely to be erroneous. Therefore we recommend requiring that at least 10 bp at each enc of the read match the reference sequence perfectly (by using the option <code>--min-read-end-match 10</code>). Additionally, the option <code>--max-repeats-in-ends</code> allows filtering reads for which the flanking region around the STR is too repetitive.
    </ul>
    </p>
    <a name="options"></a>
    <h2>Recommended allelotype options</h2>
    <p>
    Based on comparisons to capillary electrophoresis and on physical inspection of alignments for erroneous calls, we recommend the following options to <code>allelotype</code>:
    {% highlight bash %}
allelotype \
  ... \
  --filter-mapq0 \
  --filter-clipped \
  --max-repeats-in-ends 3 \
  --min-read-end-match 10
    {% endhighlight %}
    where <code>...</code> refers to required arguments to <code>allelotype</code> (see <a href="usage.html">usage</a>).

    </p>
      </section>
    </div>
    {% include footer.html %}
  </body>
</html>