---
---
<!doctype html>
<html>
  {% include header.html %}
  <body>
    <div class="wrapper">
      {% include menu.html %}
      <section>
	<h2>Changelog</h2>
	<h3>New in version 2.0.8 <em>(beta)</em></h3>
	This version should have big alignment speedups compared to previous versions, up to 5-10x faster than before. This version also fixes a minor bug fix for an off-by-one error in how allele sequences are reported in the REF and ALT fields of the VCF file.<br><br>
	<h3>New in version 2.0.5</h3>
	This version contains minor changes. Most notably we changed the way we package binaries, and the binaries should hopefully run smoothly on most systems.<br><br>
	<h3>New in version 2.0.4</h3>
	<ul>
	  This version mostly gives bug fixes:
	  <ul>
	    <li>Bug in coverage values reported in VCF files</li>
	    <li>Fixed problems parsing BAM tags in the allelotyper on BAMs that have been modified after running lobSTR.</li>
	    <li>A number of configuration cleanups that should allow lobSTR to compile on most systems.</li>
	  </ul>
	</ul>
	<h3>New in version 2.0.3</h3>
	<ul>
	  <li> Greatly expanded STR reference now includes mononucleotide repeats.</li>
	  <li> Updated VCF format automatically generated by the allelotype step.</li>
	  <li> The allelotype step can now process multiple samples at once.</li>
	  <li> Quality metrics automatically reported without the need to run additional scripts.
	  <li> Updated stutter noise scoring scheme. Full documentation is on the <a href="documentation.html">documentation page</a>.</li>
	  <li> Allow for processing paired-end bams if some mate pairs are missing.</li>
	  <li> Parameters for each run are now uploaded to Amazon S3. To turn this off, use --noweb.</li>
	  <li> Redundant aligned.tab and genotypes.tab files are no longer generated. Only BAM and VCF files are output.</li>
	  <li> Removed feature to report reads that partially span an STR </li>
	</ul>
	<h3>New in version 2.0.2</h4>
	<ul>
	  <li> Removed R dependency of allelotype step and improved stutter model.</li>
	  <li> Added option to add a read group tag to alignments, plus set map quality in BAM files to 255 in order to allow compatibility with downstream analysis tools.</li>
	  <li> Changed --sex option to --haploid, which allows users to determine which chromosomes should be treated as haploid or hemizygous. This allows for training on organisms for which the hemizygous chromosomes are not necessarily chrX and chrY.</li>
	  <li> Added the option to exclude information on partially spanning reads from the allelotype output file.
	</ul>
	
	<h3>New in version 2.0</h4>
	<ul>
	  <li>Support for paired end reads. lobSTR uses mate pair information to
	    increase confidence in STR alignments. lobSTR can still run in
	    single-end mode.</li>
	  <li>Paired-end read stitching: when possible, lobSTR will stitch
	    overlapping paired-end reads that in some cases can span longer STRs
	    than possible with a single read.
	  <li>Support for reading gzipped FASTA and FASTQ files.</li>
	  <li>(<em>beta</em>) Improved mapping quality and allelotype confidence scores</li>
	  <li>(<em>beta</em>) Alignment of partially spanning reads: lobSTR
	    reports alignments of reads that only partially span STR
	    regions. The allele length reported gives a lower bound on the true
	    STR allele.</li>
	  <li>(<em>beta</em>) Conversion of allelotype output to VCF format</li>
	  <li>(<em>beta</em>) Support for running on Amazon Web Services</li>
	</ul>

    <h2>Previous versions</h2>

    Old lobSTR indices, compatible with versions 2.0.0 and above:
    <ul>
      <li><a href="http://erlichlab.wi.mit.edu/lobSTR/lobstr_index_hg18_extend1000.tar.gz">hg18 index</a></li>
      <li><a href="http://erlichlab.wi.mit.edu/lobSTR/lobstr_index_hg19_extend1000.tar.gz">hg19 index</a></li>
    </ul>

    To download previous lobSTR versions, see <a href="http://www.broadinstitute.org/~gymrek/archive/">the lobSTR archive</a>.
    </section>
</div>
    {% include footer.html %}
  </body>
</html>
