---
---
<!doctype html>
<html>
  {% include header.html %}
  <body>
    <div class="wrapper">
      {% include menu.html %}
      <section>
	<h2> lobSTR call sets </h2>
	Below are links to VCFs we've generated using lobSTR for a variety of population-scale sequencing projects. 
	<br>
	<ul>
	  <li> lobSTR calls for Phase 1 of the 1000 Genomes Project <a href="http://erlichlab.wi.mit.edu/lobSTR/1000Genomes_Phase_1.vcf.gz"> 1000Genomes_Phase_1.vcf.gz </a> </li>
	</ul>
	<h2>lobSTR compiled binaries</h2>
	<ul>
	  <li> Linux 64 bit (x86-64) <a href="http://erlichlab.wi.mit.edu/lobSTR/lobSTR-bin-Linux-x86_64-2.0.8.tar.gz">lobSTR-bin-Linux-x86_64-2.0.8.tar.gz</a>
	</ul>
	<h2>lobSTR source code</h2>
	<ul>
	  <li><a href="http://erlichlab.wi.mit.edu/lobSTR/lobSTR-2.0.8.tar.gz">lobSTR-2.0.8.tar.gz</a>
	</ul>
	<h3>New in version 2.0.8 <em>(beta)</em></h3>
	This version should have big alignment speedups compared to previous versions, up to 5-10x faster than before. This version also fixes a minor bug fix for an off-by-one error in how allele sequences are reported in the REF and ALT fields of the VCF file.<br><br>
	To see a list of changes and download previous lobSTR versions, see the <a href="changelog.html">changelog</a> page.

    <h2>lobSTR source code on Github</h2>
    <em>*Note*: the github repository is currently under construction as we get ready to release lobSTR v3 with some major changes to our STR calling algorithm. See the README there for more details.</em>
    <br>
    The lobSTR source code is available on github (Note, we have moved over to github from sourceforge). To get the latest version, you can pull the code from the current git repo:
    <br><br>
    <tt>git clone https://github.com/mgymrek/lobstr-code.git lobstr-code</tt>
    <br><br>
    Then follow the installation instructions for building from source. 
    <p>Contact mgymrek@mit.edu if you are interested in being involved with lobSTR development.</p>
    
    <h2>lobSTR index</h2>
    The index for v2.0.3+ has been updated to expand a much broader range of STRs obtained by filtering Tandem Repeat Finder results using empirical score thresholds. Contains loci for periods 1 through 6 as well as markers obtained from CODIS, Marshfield and Y-STR datasets. More tools for how to compare markers at these annotated loci to datasets generated by other technologies is coming soon. Resource bundles contain a table of STR loci, the lobSTR index, and the file needed for the --strinfo option for the allelotyping step.
    <ul>
      <li><a href="http://erlichlab.wi.mit.edu/lobSTR/lobSTR_hg19_resource_bundle.tar.gz">hg19 resource bundle</a>
    </ul>
    <h2>lobSTR v3. index</h2>
    Starting with lobSTR v3.0.0+, we have changed our index structure. lobSTR v3 has not yet been released, but is currently available on github. If you use that version, you will need this new index. There is now one unified index, rather than files for each repeat motif. This index is not compatible with previous lobSTR versions. If you try to use the old index with v3.0.0+, you will receive an error message pointing you to where you can download the new index.
    <ul>
      <li><a href="http://files.teamerlich.org/lobstr/v3/ref/lobSTR_v3_hg19_resource_bundle.tar.gz">hg19 resource bundle for lobSTR v3</a></li>
    </ul>
    </section>
</div>
    {% include footer.html %}
  </body>
</html>
