---
---
<!doctype html>
<html>
  {% include header.html %}
  <body>
    <div class="wrapper">
      {% include menu.html %}
      <section>
	<h2>FAQ</h2>
	<strong>How do I cite lobSTR?</strong><br/>
	Please cite: <emph>Gymrek M, Golan D, Rosset S, & Erlich Y. lobSTR: A short tandem repeat profiler for personal genomes. Genome Research. 2012 April 22.</emph><br /><br />

	<strong>Can lobSTR detect STRs of periods below 2 or above 6?</strong><br/>
	The definition of what constitutes an STR is not clear cut. Currently lobSTR is limited to STR motifs of periods 1 through 6. We recently expanded our reference to include mononucleotides and a larger set of STRs. <br/><br>

	<strong>How are mismatches calculated?</strong><br />
	The parameters that control alignment mismatches/gaps are:<ul style=""><li style="">-m&nbsp;: edit distance allowed</li><li style="">-r&nbsp;: the fraction of missing alignments given 2% uniform base error rate. Note if m is set, then r is ignored.</li><li style="">-g&nbsp;: the number of gap opens allowed</li><li style="">-e&nbsp;: the number of gap extends allowed</li></ul>Each flanking region is aligned separately, so each paramter above is applied to both the left and right flanking regions. That means, if you allow an edit distance of 2 (-m 2), then edit distance 2 is allowed both in the left and right flanking sequences.<br /><br />
	
	<strong>How is the sam format generated?</strong><br />
	After an STR alignment is found, lobSTR performs a local realignment using the Needleman Wunsch algorithm with an affine gap penalty score. This step allows lobSTR to generate accurate CIGAR scores for the entire read, including the STR region, that can be used downstream for viewing in samtools tview and for calling SNPs in STR regions.<br /><br />
	
	<strong>How do you obtain the genomic coordinates of markers?</strong><br />
	We use the table of STR's generated by&nbsp;<a href="http://tandem.bu.edu/trf/trf.html" style="">Tandem Repeat Finder</a>&nbsp;available on the UCSC Genome Browser.<br /><br />
	
	<strong>How should I filter my results to obtain only high quality genotypes?</strong><br />
	lobSTR reports a score for each genotype result based on the likelihood ratios of each possible genotype, ranging from 0 to 1. We have found that this score reliable reflects the accuracy of calls. <br><br>
	
	<strong>What does the remove duplicates procedure do? How can I turn it off?&nbsp;</strong><br />
	All reads starting at the same 5' location with the same length are collapsed into a single read. If the reads have multiple copy numbers of an STR, the copy number exhibited by the majority of the PCR duplicates is retained. If there is no majority vote, quality scores are used to break ties. To turn off removing duplicates, use the --no-rmdup option to&nbsp;allelotype.<br /><br />
	
	<strong>How can I turn off trimming?</strong><br />
	lobSTR trims reads based on quality scores to remove erroneous read
	ends. The --bwaq acts the same as the -q parameter for
	BWA. Additionally, by default, lobSTR trims all flanking regions to 25bp before aligning. To avoid this step, set --maxflank to a value near the read length of the input.<br /><br/>
	
	<strong>What does the --noise_model option take?</strong><br />
	The --noise_model option takes a file name to read or write the noise model parameters to. If you specify --command classify, the allelotype program reads in the noise model file given and uses those parameters. If you specify --command train, allelotype will write to the file given to noise_model with the new trained parameters. Note that training is only reliable when the sample is male and contains a large number of reads aligned to sex chromosomes, so in most cases we recommend just using the --command classify option with a provided noise model.<br /><br/>
	
	<strong>The lobSTR indexer did not produce any output.</strong><br />
	In versions below 2.0.0, lobSTR requires chromosome names to not
	contain "_". If lobSTR does not return any output from the indexing
	step, it may be because this illegal character is present. This has
	been changed in version 2.0.0 to allow "_" but disallow "$" in
	chromosome names.<br/><br/>
	
	<strong>What read lengths can lobSTR align?</strong><br />
	Because most STRs in lobSTR's reference are ~40bp, and a read must
	fully span an STR plus some flanking region in order to produce a
	unique alignment, lobSTR works best with reads of 100bp or more. Some high
	quality alignments can be produced with reads as short as 45bp, but alignment errors are much more frequent with these short reads. 36bp
	reads cannot be reliably aligned by lobSTR.
	<br/><br/>
      </section>
    </div>
    {% include footer.html %}
  </body>
</html>
