---
---
<!doctype html>
<html>
  {% include header.html %}
  <body>
    <div class="wrapper">
      {% include menu.html %}
      <section>
	<h2>Output file formats (as of v3.0.2)<br /></h2>

	<h3>$prefix.aligned.bam</h3>
	A bam file containing all aligned
	reads. This file contains the same information as the aligned.tab
	file. This is in the standard bam format, with the following added
	fields:
	<ul>
	  <li>XS: start position of matching STR</li>
	  <li>XE: end position of matching STR</li>
	  <li>XR: STR repeat</li>
	  <li>XC: reference copy number</li>
	  <li>XD: nucleotide length difference compared to reference</li>
	  <li>XX: 0/1 flag to indicate stitching of paired end reads</li>
	  <li>XM: distance between mate pairs.</li>
	  <li>XN: (only if given) name of the STR repeat</li>
	  <li>XG: The extracted STR region</li>
	  <li>XQ: map quality score of the aligned read</li>
	  <li>RG: (only if given) read group</li>
	  <li>XA: alternate alignments for multimappers</li>
	  <li>XO: other STRs spanned by this read</li>
	</ul>
	Note the bam output is not sorted.<br /><br />
	
	<h3>$prefix.vcf</h3>
	<h4>Required fields</h4>
	<ul>
	  <li> CHROM: chromosome of the STR</li>
	  <li> POS: start position of the STR. Note, since the start position is
	    used, STRs with the same start coordinate but different motifs must
	    be filtered for compatibility with downstream vcftools.</li>
	  <li> ID: this field is set to ".". For annotated STR loci, this gives the name of the STR marker.</li>
	  <li> REF: gives the base in the reference at CHROM:POS.</li>
	  <li> ALT: contains a comma separated list of "&#60;STRVAR:allele&#62;" where allele is the number of base pairs different from reference. If any STR variations were reported.
	  <li> QUAL: contains -10log10(P), where P is the probability that all samples are homozygous reference.</li>
	  <li> FILTER: this field is left blank</li>
	</ul>
	
	<h4>INFO fields</h4>
	<ul>
	  <li> RPA: repeats per allele</li>
	  <li> RL: total length of the STR tract in the reference</li>
	  <li> REF: the reference copy number </li>
	  <li> END: the end coordinate of the STR</li>
	  <li> MOTIF: the canonicalized STR repeat motif </li>
	  <li> RU: repeat motif as it appears on the forward strand in the reference</li>
	  <li> VT: variant type, set to STR </li>
	</ul>
	<h4>FORMAT fields</h4>
	<ul>
	  <li> ALLREADS: string giving all alleles of all reads seen (column 11 from genotypes.tab
	    file). The format is allele1|readcount;allele2|readcount, etc.</li>
	  <li> DP: STR coverage</li>
	  <li> GB: reported allelotype given in bp difference from reference. Set to "./." for no call.</li>
	  <li> Q: Likelihood ratio score of the allelotype call. Gives the likelihood of the reported genotype divided by the sum of likelihoods of all considered genotypes.</li>
	  <li> PL: Phred-scaled genotype likelihoods. Given for each possible pair of alleles from the ALT field. If j and k are the indices of the alleles in the ALT field, the (F(j/k) = (k*(k+1)/2)+j)th field gives the likelihood of allelotype (allele j, allele k) as in the standard VCF format. Normalized to give likelihoods relative to that of the maximum likelihood genotype, which will have a phred-scaled likelihood of 0.</li>
	  <li> AML: allele marginal likelihood ratio scores. The score for each allele gives the sum of all likelihoods of genotypes containing that allele divided by the sum of likelihoods of all genotypes considered.</li>
	  <li> GT: genotype, given as indices into the list in the ALT field as
	    in the standard VCF format. Set to "./." if no call.</li>
	  <li> STITCH: Number of stitched reads</li>
	</ul>
      </section>
    </div>
    {% include footer.html %}
  </body>
</html>
