---
---
<!doctype html>
<html>
  {% include header.html %}
  <body>
    <div class="wrapper">
      {% include menu.html %}
      <section>
	<h2>Output file formats (as of v4.0.0)<br /></h2>

	<h3><em>PREFIX</em>.aligned.bam</h3>
	A BAM file containing all aligned
	reads. This is in the standard BAM format, with the following added
	tags:
	<ul>
	  <li>XS: start position of matching STR. Tag type: i</li>
	  <li>XE: end position of matching STR. Tag type: i</li>
	  <li>XR: STR repeat. Tag type: Z</li>
	  <li>XC: reference copy number. Tag type: f</li>
	  <li>XD: nucleotide length difference compared to reference. Tag type: i</li>
	  <li>XX: 0/1 flag to indicate stitching of paired end reads. Tag type: i</li>
	  <li>XM: distance between mate pairs. Tag type: i</li>
	  <li>XN: (only if given) name of the STR repeat. Tag type: Z</li>
	  <li>XG: The extracted STR region. Tag type: z</li>
	  <li>XQ: map quality score of the aligned read. Tag type: i</li>
	  <li>RG: read group. Tag type: Z</li>
	  <li>XA: alternate alignments for multimappers. Tag type: Z</li>
	  <li>XO: other STRs spanned by this read. Tag type: Z</li>
	</ul>
	Note the BAM output is not sorted.<br /><br />
	
	<h3><em>PREFIX</em>.vcf</h3>
	<h4>Required fields</h4>
	<ul>
	  <li> CHROM: chromosome of the STR</li>
	  <li> POS: start position of the STR. Note, since the start position is
	    used, STRs with the same start coordinate but different motifs must
	    be filtered for compatibility with downstream vcftools.</li>
	  <li> ID: this field is set to ".". For annotated STR loci, this gives the name of the STR marker.</li>
	  <li> REF: gives the base in the reference at CHROM:POS.</li>
	  <li> ALT: contains a comma separated list of "&#60;STRVAR:allele&#62;" where allele is the number of base pairs different from reference. If any STR variations were reported.
	  <li> QUAL: contains -10log10(P), where P is the probability that all samples are homozygous reference.</li>
	  <li> FILTER: this field is left blank</li>
	</ul>
	
	<h4>INFO fields</h4>
	<ul>
	  <li> RPA: repeats per allele</li>
	  <li> END: the end coordinate of the STR</li>
	  <li> MOTIF: the canonicalized STR repeat motif </li>
	  <li> NS: Number of samples with data </li>
	  <li> REF: the reference copy number </li>
	  <li> RL: total length of the STR tract in the reference</li>
	  <li> RU: repeat motif as it appears on the forward strand in the reference</li>
	  <li> VT: variant type, set to STR </li>
	</ul>
	<h4>FORMAT fields</h4>
	<ul>
	  <li> ALLREADS: string giving all alleles of all reads seen (column 11 from genotypes.tab
	    file). The format is allele1|readcount;allele2|readcount, etc.</li>
	  <li> AML: allele marginal likelihood ratio scores. The score for each allele gives the sum of all likelihoods of genotypes containing that allele divided by the sum of likelihoods of all genotypes considered.</li>
	  <li> DISTENDS: Average difference between distance of STR to read ends.</li>
	  <li> DP: STR coverage</li>
	  <li> GB: reported allelotype given in bp difference from reference. Set to "./." for no call.</li>
	  <li> PL: Phred-scaled genotype likelihoods. Given for each possible pair of alleles from the ALT field. If j and k are the indices of the alleles in the ALT field, the (F(j/k) = (k*(k+1)/2)+j)th field gives the likelihood of allelotype (allele j, allele k) as in the standard VCF format. Normalized to give likelihoods relative to that of the maximum likelihood genotype, which will have a phred-scaled likelihood of 0.</li>
	  <li> Q: Likelihood ratio score of the allelotype call. Gives the likelihood of the reported genotype divided by the sum of likelihoods of all considered genotypes.</li>
	  <li> GT: genotype, given as indices into the list in the ALT field as
	    in the standard VCF format. Set to "./." if no call.</li>
	  <li> SB: strand bias (as defined in the <a href="https://www.broadinstitute.org/gatk/guide/tooldocs/org_broadinstitute_gatk_tools_walkers_annotator_StrandOddsRatio.php">GATK documentation</a></li>
	  <li> STITCH: Number of stitched reads</li>
	  <li> FT: (filled by <code>lobSTR_filter_vcf.py</code>) Call-level filter.</li>
	</ul>
      </section>
    </div>
    {% include footer.html %}
  </body>
</html>
