---
---
<!doctype html>
<html>
  {% include header.html %}
  <body>
    <div class="wrapper">
      {% include menu.html %}
      <section>
      <h2>Filtering lobSTR VCFs</h2>
      <ul>
      <li><a href="#overview">Overview</a></li>
      <li><a href="#filter">Using <code>lobSTR_filter_vcf.py</code></a>
        <ul>
          <li><a href="#basic">Basic usage</a></li>
          <li><a href="#output">Output files</a></li>
          <li><a href="#cl">All command line arguments</a></li>
        </ul>
      </li>
      <li><a href="#recommended">Recommended filters</a></li>
      </ul>
  <p>
    <a name="overview"></a>
    <h2>Overview</h2>
    This page describes how to use <code>lobSTR_filter_vcf.py</code> to filter VCF files generated by lobSTR based on a number of metrics. It then describes our recommended filters for multi-sample calling. Note this script will only work with VCFs generated by lobSTR v4.0.0+.
  </p>
    <p>
    <a name="filter"></a>
    <h2>Using <code>lobSTR_filter_vcf.py</code></h2>
    <p>
    If you followed the instructions on the <a href="install.html">install</a> page, <code>lobSTR_filter_vcf.py</code> should be installed in <code>$PREFIX/share/lobSTR/scripts/lobSTR_filter_vcf.py</code>. This tool applies locus and call level filters which are annotated in the output VCF file.
    </p>
    <a name="basic"></a>
    <h3>Basic usage</h3>
    <p>
      To run <code>lobSTR_filter_vcf.py</code>:
{% highlight bash %}
lobSTR_filter_vcf.py --vcf VCFFILE > OUTFILE
{% endhighlight %}
where VCFFILE is a VCF file generated by lobSTR and OUTFILE is an output VCF file. By default no filters are applied. A number of options allow for specific locus and call level filters. All locus level filters are of the form <code>--loc-XX</code> and all call level filters are of the form <code>--call-XX</code>. For example:
{% highlight bash %}
lobSTR_filter_vcf.py --vcf VCFFILE --loc-max-ref-length 80 --call-cov 10 > OUTFILE
{% endhighlight %}
will filter all loci with a reference length of greater than 80bp and all calls with less than 10x coverage.
    </p>
    <a name="output"></a>
    <h3>Output files</code></h2>
    <p>
    The output file from this script is in standard VCF format. Loci remaining after filtering will have "PASS" in the FILTER column of the VCF. Non-passing loci will have a comma-separated list of filters that failed. If it was not present, an "FT" field will be added to the FORMAT field. This is a string field that will say:
    <ul>
    <li>"PASS" for all calls that passed filters.</li>
    <li>"NOCALL" for loci with no call information</li>
    <li>A ":" separated list of filters that failed if the call did not pass all filters.</li>
    </ul>
    </p>
    <a name="cl"></a>
    <h3>All command line arguments</h3>
    <p>
    The table below lists filtering options that are currently available. If you would like additional filter options added, feel free to <a href="contact-us.html">contact us</a> or submit a pull request on <a href="https://github.com/mgymrek/lobstr-code/">github</a>.
  <table class="box-table-a">
    <tr><th>Parameter</th><th>Description</th><th>Type</th><th>Default</th></tr>
    <tr>
      <td><code>--vcf</code></td>
      <td>(REQUIRED) Input VCF file (generated by <code>allelotype)</code></td>
      <td>string</td>
      <td>-</td>
    </tr>
    <tr>
      <td><code>--loc-log-score</code></td>
      <td>Min mean -log10(1-Q) cutoff to include a locus</td>
      <td>float</td>
      <td>0.0</td>
    </tr>
    <tr>
      <td><code>--loc-cov</code></td>
      <td>Min mean coverage to include a locus.</td>
      <td>int</td>
      <td>0</td>
    </tr>
    <tr>
      <td><code>--loc-max-ref-length</code></td>
      <td>Max reference length of a locus to include</td>
      <td>int</td>
      <td>10000</td>
    </tr>
    <tr>
      <td><code>--loc-call-rate</code></td>
      <td>Min call rate to include a locus.</td>
      <td>float</td>
      <td>0.0</td>
    </tr>
    <tr>
      <td><code>--call-dist-end</code></td>
      <td>Max mean absolute difference in distance from read ends to include a call.</td>
      <td>float</td>
      <td>100.0</td>
    </tr>
    <tr>
      <td><code>--call-log-score</code></td>
      <td>Min mean -log10(1-Q) cutoff to include a call</td>
      <td>float</td>
      <td>0.0</td>
    </tr>
    <tr>
      <td><code>--call-cov</code></td>
      <td>Min coverage to include a call</td>
      <td>int</td>
      <td>0</td>
    </tr>
    <tr>
      <td><code>--ignore-samples</code></td>
      <td>Ignore these samples when appying filters. File with one sample/line.</td>
      <td>string</td>
      <td>-</td>
    </tr>
  </table>
    </p>
  </p>
  </p>
    <p>
    <a name="recommended"></a>
    <h2>Recommended filters</code></h2>
    Based on comparison with capillary electrophoresis data, we recommend the follow quality filters for high coverage whole genome sequencing data:
    <ul>
      <li>Average coverage at least 5x (<code>--loc-cov 5</code>)</li>
      <li>Average -log10(1-Q) at least 0.8 (<code>--loc-log-score 0.8</code>)</li>
      <li>Call rate at least 80% (<code>--loc-call-rate 0.8</code>)</li>
      <li>Reference allele length less than or equal to 80bp (<code>--loc-max-ref-length 80</code>)</li>
      <li>Call coverage at least 5x (<code>--call-cov 5</code>)</li>
      <li>Call -log10(1-Q) at least 0.8 (<code>--call-log-score 0.8</code>)</li>
      <li>Absolute value of DISTENDS score at most 20 (<code>--call-dist-ends 20</code>)</li>
    </ul>
    Alternative filters may be appropriate for other types of datasets.
  </p>
      </section>
    </div>
    {% include footer.html %}
  </body>
</html>