---
---
<!doctype html>
<html>
  {% include header.html %}
  <body>
    <div class="wrapper">
      {% include menu.html %}
      <section>
	<h2>Overview</h2>
	<ul>
	  <li><a href="#binary">Installation from precompiled binaries</a></li>
	  <li><a href="#source">Installation from source</a></li>
	  <li><a href="#osx">Installation on Mac OSX</a></li>
	  <li><a href="#testinstall">Testing that lobSTR was installed correctly</a></li>
	  <li><a href="#req">Requirements</a></li>
	</ul>
	<a name="binary"></a>
	<h2>Installation from precompiled binaries</h2>
	{% highlight bash %}
wget http://files.teamerlich.org/lobSTR/lobSTR-bin-Linux-x86_64-X.X.X.tar.gz
tar -xzf lobSTR-bin-Linux-x86_64-X.X.X.tar.gz
cd lobSTR-bin-Linux-x86_64-X.X.X
sudo cp bin/lobSTR bin/allelotype bin/lobSTRIndex /usr/local/bin
	{% endhighlight %}
	<p>
	  <a name="source"></a>
	  <h2>Installation from source</h2>
	  Before installing from source, make sure all required packages are installed and in the user's $PATH (see below for a list of requirements). lobSTR should compile on most Linux and Mac systems.
	  {% highlight bash %}
wget http://files.teamerlich.org/lobSTR/lobSTR-X.X.X.tar.gz
tar -xzf lobSTR-X.X.X.tar.gz
cd lobSTR-X.X.X/
./configure
make
make check
sudo make install
	  {% endhighlight %}
	  <p>
	  You can find the latest version of the source code on the <a href="https://github.com/mgymrek/lobstr-code">git repository</a>.
	  </p>
	<p>
	  Scripts will be available in <code>$PREFIX/share/lobSTR/scripts</code> and noise model files in <code>$PREFIX/share/lobSTR/models</code>. $PREFIX is by default <code>/usr/local</code> on Linux systems unless you specified a different prefix during the installation.
	</p>
	<p>
	For usage instructions see&nbsp;<a href="usage.html">usage</a>.
	</p>
	<a name="osx"></a>
	<h2>Installation on Mac OSX</h2>
	<p>
	You can also install lobSTR using <a href="http://brew.sh/">Homebrew</a>. First make sure homebrew is installed, then run:
	{% highlight bash %}
brew tap homebrew/science
brew install lobstr
	{% endhighlight %}
	</p>
	<a name="testinstall"></a>
	<h2>Testing that lobSTR was installed correctly</h2>
	<p>
	To test that your lobSTR installation is working, you can run the provided test script installed to <code>$PREFIX/share/lobSTR/scripts/lobSTR_test_run.sh</code>. $PREFIX is by default <code>/usr/local/</code>. If this runs without errors, lobSTR was installed successfully.
	</p>
	<a name="req"></a>
	<h2>Requirements</h2>
	<p>
	The packages below are required for building lobSTR from source. If you are using the precompiled binaries, these are not required.
	</p>
	<strong><font color="#660000" size="3">External libraries</font></strong>
	<table cellspacing="15">
	  <tr>
	  <td class="table_hdr">Program</td>
	  <td class="table_hdr">Tested on</td>
	  <td class="table_hdr">Debian/Ubuntu package</td>
	  <td>Homebrew package</td>
	  </tr>
	  <tr>
	  <td class="table_entry">gcc</td>
	  <td class="table_entry">4.4.3</td>
	  <td class="table_entry">gcc</td>
	  <td class="table_entry">gcc</td>
	  </tr>
	  <tr>
	  <td class="table_entry">g++</td>
	  <td class="table_entry">4.4.3</td>
	  <td class="table_entry">g++</td>
	  <td class="table_entry">g++</td>
	  </tr>
	  <tr>
	  <td class="table_entry">automake</td>
	  <td class="table_entry">1.11.1</td>
	  <td class="table_entry">automake</td>
	  <td class="table_entry">automake</td>
	  </tr>
	  <tr>
	  <td class="table_entry">libtool</td>
	  <td class="table_entry">2.2.6</td>
	  <td class="table_entry">libtool</td>
	  <td class="table_entry">libtool</td>
	  </tr>
	  <tr>
	  <td class="table_entry">pkg-config</td>
	  <td class="table_entry">0.22</td>
	  <td class="table_entry">pkg-config</td>
	  <td class="table_entry">pkg-config</td>
	  </tr>
	  <tr>
	  <td class="table_entry">boost</td>
	  <td class="table_entry">1.46.1</td>
	  <td class="table_entry">libboost-dev</td>
	  <td class="table_entry">boost</td>
	  </tr>
	  <tr>
	  <td class="table_entry">zlib</td>
	  <td class="table_entry">1.2.3.3</td>
	  <td class="table_entry">zlib1g-dev</td>
	  <td class="table_entry">-</td>
	  </tr>
	  <tr>
	  <td class="table_entry">cppunit</td>
	  <td class="table_entry">1.12.1</td>
	  <td class="table_entry">libcppunit-dev</td>
	  <td class="table_entry">cppunit</td>
	  </tr>
	  <tr>
	  <td class="table_entry">gsl</td>
	  <td class="table_entry">1.16</td>
	  <td class="table_entry">libgsl10-dev</td>
	  <td class="table_entry">gsl</td>
	  </tr>
	</table>
	<font color="#660000" size="3"><strong>Python
	    libraries</strong></font><br />Note: the python libraries listed below are only required for certain optional steps. All Python scripts were tested using Python 2.7.
	<table cellspacing="15">
	 	<tr>
	  	<td class="table_hdr">Program</td>
	  	<td class="table_hdr">Tested on</td>
	  	<td class="table_hdr">Download link</td>
	  	<td class="table_hdr">easy_install package</td>
	  	<td class="table_hdr">lobSTR steps where required</td>
	  	</tr>
	  	<tr>
	  	<td class="table_entry">numpy</td>
	  	<td class="table_entry">1.8.2</td>
	  	<td class="table_entry"><a href="http://www.numpy.org/">download</a></td>
	  	<td class="table_entry">numpy</td>
	  	<td class="table_entry">Filtering VCFs</td>
	  	</tr>
	  	<tr>
	  	<td class="table_entry">pyfasta</td>
	  	<td class="table_entry">0.5.1</td>
	  	<td class="table_entry"><a href="https://pypi.python.org/pypi/pyfasta/">download</a></td>
	  	<td>pyfasta</td><td>Building a custom index and building an strinfo file required for allelotyping from a custom index for v3.*.</td>
	  	</tr>
	  	<tr>
	  	<td class="table_entry">pyvcf</td>
	  	<td class="table_entry">0.6.7</td>
	  	<td class="table_entry"><a href="http://pyvcf.readthedocs.org/en/latest/index.html">download</a></td>
	  	<td class="table_entry">pyvcf</td>
	  	<td class="table_entry">Filtering VCFs, vcf-extract, vcf-to-tab</td>
	  	</tr>
	  	<tr>
	  	<td class="table_entry">Biopython</td>
	  	<td class="table_entry">1.5.6</td>
	  	<td class="table_entry"><a href="http://biopython.org/wiki/Download">download</a></td>
	  	<td>biopython</td>
	  	<td>Building a custom index and building an strinfo file required for allelotyping from a custom index for v2.*.</td>
	  	</tr>
	</table>
      </section>
    </div>
    {% include footer.html %}
  </body>
</html>
