---
---
<!doctype html>
<html>
  {% include header.html %}
  <body>
    <div class="wrapper">
      {% include menu.html %}
      <section>
	<h2>lobSTR Reference and Index</h2>
	lobSTR aligns reads specifically to places in the reference genome known to contain short tandem repeats. Based on a reference set of STRs, lobSTR builds a special index data structure to allow it to efficiently align reads to STR-containing regions. The structure of the reference BED file is described on the <a href="http://lobstr.teamerlich.org/usage.html">usage page</a> under "Building a custom index". The sections below describe the structure of the lobSTR index created from a reference set of STRs.
	<ul>
	  <li><a href="#index2">Index Structure: Version 2</a></li>
	  <li><a href="#index3">Index Structure: Version 3</a></li>
	  <li><a href="#changelog">Changelog</a></li>
	</ul>
	<p>
	  <a name="index2">
	<h3>Index Structure: Version 2</h3>
	This section describes the index structure used for lobSTR v2.0.3+. A new index structure was introduced with lobSTR v3+ (see below). The lobSTR index consists of a collection of files within a single directory with the same prefix ("lobSTR_" by default). It contains the following files:
	  <ol>
	    <li>A separate <a href="http://bio-bwa.sourceforge.net/bwa.shtml">BWA</a>-style index for each class of repeat unit. This is made up of the following files for each repeat unit:
	    <ul>
	      <li>lobSTR_$repeat.amb</li>
	      <li>lobSTR_$repeat.ann</li>
	      <li>lobSTR_$repeat.bwt</li>
	      <li>lobSTR_$repeat.pac</li>
	      <li>lobSTR_$repeat.rbwt</li>
	      <li>lobSTR_$repeat.rpac</li>
	      <li>lobSTR_$repeat.rsa</li>
	      <li>lobSTR_$repeat.sa</li>
	      <li>lobSTR_$repeat.fa</li>
	    </ul>
	    These file types are generated using BWA.
	    </li>
	    <li>
	      <b>lobSTR_ref.fa</b>: This file contains one entry per STR in the reference genome. The sequence for each entry consists of the STR +/- 1kb of flanking region. The header consists of the following fields separated by "$"'s:
	      <ul>
		<li>ID number of the entry</li>
		<li>Chromosome</li>
		<li>Start coordinate of the region in the original reference genome</li>
		<li>End coordinate of the region in the original reference genome</li>
		<li>Canonicalized repeat motif</li>
		<li>Copy number in the reference genome</li>
		<li>Repeat motif in the reference genome</li>
	      </ul>
	    </li>
	    <li>
	      <b>lobSTR_strdict.txt</b>: This file contains additional metadata about the index used by lobSTR. The head of the file gives the size of each chromosome in the original reference genome. Below the header lists each repeat unit for which there is a reference index.
	    </li>
	  </ol>
	  In total, the hg19 reference download consists of 4,115 files: 9 BWA index files for each 457 repeats, plus lobSTR_ref.fasta and lobSTR_strdict.txt.
	<p>
	  <a name="index3">
	<h3>Index Structure: Version 3</h3>

	<p>
	  <a name="changelog">
	<h3>Changelog</h3>
      </section>
    </div>
    {% include footer.html %}
  </body>
</html>
