---
---
<!doctype html>
<html>
  {% include header.html %}
  <body>
    <div class="wrapper">
      {% include menu.html %}
      <section>
	<h2>You will need</h2>
	<ul>
	  <li>Download the <a href="download.html">lobSTR</a> package.</li>
	  <li>Download the lobSTR resource bundle for the appropriate reference genome (available on the <a href="download.html">downloads</a> page). If you'd like to use your own custom set of STR loci, see <a href="best-practices-custom-reference.html">Building a custom reference</a> before proceeding.</li>
	</ul>
	<h2>Announcements</h2>
	<ul>
	<li>
		<strong>As of v4+, <code>allelotype</code> can now take input BAM files generated by aligners other than <code>lobSTR</code></strong>. The only aligner that has been tested and shown to work well is <a href="http://bio-bwa.sourceforge.net/">BWA-MEM</a>. For more details on running <code>allelotype</code> with <code>BWA-MEM</code>-generated BAMs as input, see <a href="best-practices-bwamem.html">Best practices for using lobSTR with BWA-MEM alignments</a>
	</li>
	<li>
		Using noise models generated by previous <code>allelotype</code> versions (before v3.0.0) with versions 3.0.0 and above will result in erroneous genotype calls. If using <code>allelotype</code> 3.0.0+, <strong>you must use the new noise model (<code>illumina.v3.pcrfree</code>) or a custom noise model built with the new version.</strong>
	</li>
	</ul>
	<h2>On this page</h2>
	<ul>
	  <li><a href="#basic">Basic usage</a></li>
	  <li><a href="#arvados">Running lobSTR on Curoverse</a></li>
	  <li><a href="#detailed">Detailed description of all command line options</a>
	    <ul>
	      <li><a href="#lobstr">lobSTR</a></li>
	      <li><a href="#allelotype">allelotype</a></li>
	      <li><a href="#index">index</a></li>
	    </ul>
	  </li>
	</ul>
	Only basic lobSTR usage is described here. Documentation for additional scripts and more usage details can be found on the <a href="documentation.html">documentation</a> page.
	<p>
	  <a name="basic"></a>
	  <h2>Basic usage (as of v4.0.0)</h2>
	<strong>To run lobSTR using default parameters</strong>:
	<p>
	Single-end reads:
	{% highlight bash %}
lobSTR -f FILE1,FILE2,.. \
   --index-prefix PATH_TO_INDEX/lobSTR_ \
   -o OUTPUT_PREFIX \
   --rg-sample SAMPLE \
   --rg-lib LIB
	{% endhighlight %}
	Paired-end reads:
	{% highlight bash %}
lobSTR --p1 FILE1,FILE2,.. --p2 FILE1,FILE2,...	\
   --index-prefix PATH_TO_INDEX/lobSTR_ \
   -o OUTPUT_PREFIX
	{% endhighlight %}
	Sort and index:
	{% highlight bash %}
samtools sort OUTPUT_PREFIX.aligned.bam OUTPUT_PREFIX.sorted
samtools index OUTPUT_PREFIX.sorted.bam
	{% endhighlight %}
	Run allelotype:
	{% highlight bash %}
allelotype \
  --command [train|classify] \
  --bam OUTPUT_PREFIX.sorted.bam \
  --noise_model NOISEMODELPREFIX \
  --out OUTPUT_PREFIX \
  --strinfo STRINFOFILE \
  --index-prefix PATH_TO_INDEX/lobSTR_
	{% endhighlight %}
	<b>To run an example using files available in the download</b>:
	{% highlight bash %}
lobSTR \
  --p1 PATH_TO_LOBSTR/tests/tmp_1.fq \
  --p2 PATH_TO_LOBSTR/tests/tmp_2.fq \
  -q \
  --index-prefix hg19_v3.0.2/lobstr_v3.0.2_hg19_ref/lobSTR_ \
  -o test \
  --rg-sample mysample --rg-lib mylibrary

samtools sort test.aligned.bam test.sorted
samtools index test.sorted.bam

allelotype \
  --command classify \
  --bam test.sorted.bam \
  --noise_model PATH_TO_LOBSTR/models/illumina_v3.pcrfree \
  --out test \
  --strinfo hg19_v3.0.2/lobstr_v3.0.2_hg19_strinfo.tab \
  --index-prefix hg19_v3.0.2/lobstr_v3.0.2_hg19_ref/lobSTR_
	{% endhighlight %}

	<p>
	The index and strinfo files are available in the resource bundle that can be found on the <a href="download.html">downloads page</a>.
	</p>
	<p>
	  Note that the arguments <code>--rg-sample</code>
	  and <code>--rg-lib</code> are now required to run lobSTR. These
	  flags add the appropriate read group information to the
	  output bam file which is important for <a href="best-practices-multi-sample-calling.html">multi-sample calling</a>. For more information on these, see
	  the <a href="faq.html">faq page</a>.
	</p>
	This will create the example output files:
	<ul>
	  <li><code>test.aligned.bam</code></li>
	  <li><code>test.aligned.stats</code></li>
	  <li><code>test.vcf</code></li>
	  <li><code>test.allelotype.stats</code></li>
	</ul>
	To see a description of the output file formats, see <a href="file-formats.html">file formats</a>. To see all available parameters, run <code>lobSTR --help</code> and <code>allelotype --help</code>.<br />
	<br><br>
	<a name="arvados"></a>
	<h2>Running lobSTR on Curoverse</h2>
	You can now run the lobSTR v3 pipeline on Curoverse, which is built with the <a href="https://arvados.org/">Arvados</a> open source project and makes it easy to manage data and pipelines. To create a free account, go to <a href="https://curoverse.com">https://curoverse.com</a> and click "Log In." A step-by-step tutorial for running lobSTR v3 on Curoverse is posted <a href="https://arvados.org/projects/arvados/wiki/LobSTR_tutorial">here</a>. If you run into any difficulties or have any feature suggestions, email support@curoverse.com or join the #arvados channel on IRC.
	<p>
	<a name="detailed"></a>
	<h2>Detailed description of command line options</h2>
	<a name="lobstr"></a>
	<h3>lobSTR</h3>
	<table class="box-table-a">
	  <tr><th>Parameter</th><th>Description</th><th>Type</th><th>Default</th></tr>
	  <tr>
	    <td><code>-f<br>--files</code></td>
	    <td>file or comma-separated list of single ends in fasta, fastq, or BAM files to align. Also use this parameter to specify
	    paired-end BAM files to align.</td>
	    <td>string</td>
	    <td>-</td>
	  </tr>
	  <tr>
	    <td><code>--p1</code></td>
	    <td>file or comma-separated list of files
	    containing the first end of paired end reads in fasta
	    or fastq format</td>
	    <td>string</td>
	    <td>-</td>
	  </tr>
	  <tr>
	    <td><code>--p2</code></td>
	    <td>file or comma-separated list of files
	    containing the second end of paired end reads in fasta
	    or fastq format</td>
	    <td>string</td>
	    <td>-</td>
	  </tr>
	  <tr>
	    <td><code>-o<br>--out</code></td>
	    <td>(REQUIRED) prefix for output files. will output <code>prefix.aligned.bam</code> and <code>prefix.aligned.stats</code>.</td>
	    <td>string</td>
	    <td>-</td>
	  </tr>
	  <tr>
	    <td><code>--index-prefix</code></td>
	    <td>(REQUIRED) prefix for lobSTR's bwa reference (must run lobstr_index.py to create index or download from lobSTR website). If the index is in PATH_TO_INDEX, this argument is PATH_TO_INDEX/lobSTR_.</td>
	    <td>string</td>
	    <td>-</td>
	  </tr>
	  <tr>
	    <td><code>--rg-sample</code></td>
	    <td>(REQUIRED) Use this in the read group SM tag</td>
	    <td>string</td>
	    <td>-</td>
	  </tr>
	  <tr>
	    <td><code>--rg-lib</code></td>
	    <td>(REQUIRED) Use this in the read group LB tag</td>
	    <td>string</td>
	    <td>-</td>
	  </tr>
	  <tr>
	    <td><code>-h<br>--help</code></td>
	    <td>display help screen</td>
	    <td>bool</td>
	    <td>-</td>
	  </tr>
	  <tr>
	    <td><code>-v<br>--verbose</code></td>
	    <td>print out useful progress messages</td>
	    <td>bool</td>
	    <td>-</td>
	  </tr>
	  <tr>
	    <td><code>--quiet</code></td>
	    <td>don't print anything to stderr or stdout</td>
	    <td>bool</td>
	    <td>-</td>
	  </tr>
	  <tr>
	    <td><code>--version</code></td>
	    <td>print out lobSTR program version</td>
	    <td>bool</td>
	    <td>-</td>
	  </tr>
	  <tr>
	    <td><code>-q<br>--fastq</code></td>
	    <td>reads are in fastq format</td>
	    <td>bool</td>
	    <td>-</td>
	  </tr>
	  <tr>
	    <td><code>--bam</code></td>
	    <td>reads are in single-end BAM format</td>
	    <td>bool</td>
	    <td>-</td>
	  </tr>
	  <tr>
	    <td><code>--gzip</code></td>
	    <td>the input files are gzipped (fastq or fasta input)</td>
	    <td>bool</td>
	    <td>-</td>
	  </tr>
	  <tr>
	    <td><code>--bampair</code></td>
	    <td>Reads are in BAM format and are paired-end. NOTE: BAM file MUST be sorted or collated by read name. (e.g. samtools sort -n file.bam prefix)</td>
	    <td>bool</td>
	    <td>-</td>
	  </tr>
	  <tr>
	    <td><code>--bwaq</code></td>
	    <td>trim reads based on quality score. Same as
	    the -q parameter in BWA</td>
	    <td>int</td>
	    <td>10</td>
	  </tr>
	  <tr>
	    <td><code>--oldillumina</code></td>
	    <td>specifies that base pair quality scores
	    are reported in the old Phred format (Illumina 1.3+,
	    Illumina 1.5+) where quality scores are given as Phred
	    + 64 rather than Phred + 33</td>
	    <td>bool</td>
	    <td>-</td>
	  </tr>
	  <tr>
	    <td><code>--multi</code></td>
	    <td>Report reads mapping to multiple genomic locations</td>
	    <td>bool</td>
	    <td>-</td>
	  </tr>
	  <tr>
	    <td><code>--noweb [deprecated]</code></td>
	    <td>Do not report any user information to Amazon S3. [This feature has been discontinued]</td>
	    <td>bool</td>
	    <td>-</td>
	  </tr>
	  <tr>
	    <td><code>-p</code></td>
	    <td>number of threads to use</td>
	    <td>int</td>
	    <td>1</td>
	  </tr>
	  <tr>
	    <td><code>--min-read-length</code></td>
	    <td>don't process reads shorter than this</td>
	    <td>int</td>
	    <td>45</td>
	  </tr>
	  <tr>
	    <td><code>--max-read-length</code></td>
	    <td>don't process reads longer than this</td>
	    <td>int</td>
	    <td>1024</td>
	  </tr>
	  <tr>
	    <td><code>--fft-window-size</code></td>
	    <td>size of sliding entropy window</td>
	    <td>int</td>
	    <td>16</td>
	  </tr>
	  <tr>
	    <td><code>--fft-window-step</code></td>
	    <td>step size of the sliding window</td>
	    <td>int</td>
	    <td>4</td>
	  </tr>
	  <tr>
	    <td><code>--entropy-threshold</code></td>
	    <td>threshold score to call a window periodic</td>
	    <td>float</td>
	    <td>0.45</td>
	  </tr>
	  <tr>
	    <td><code>--minflank</code></td>
	    <td>minimum length of flanking region to try to align</td>
	    <td>int</td>
	    <td>8</td>
	  </tr>
	  <tr>
	    <td><code>--maxflank</code></td>
	    <td>length to trim flanking regions to before aligning</td>
	    <td>int</td>
	    <td>100</td>
	  </tr>
	  <tr>
	    <td><code>--max-diff-ref</code></td>
	    <td>only report reads differing by at most this number of bp from the reference allele</td>
	    <td>int</td>
	    <td>50</td>
	  </tr>
	  <tr>
	    <td><code>--extend</code></td>
	    <td>Number of bp the reference was extended
                           when building the index.
                           Must be same as --extend parameter used
                           to run lobstr_index.py</td>
	    <td>int</td>
	    <td>1000</td>
	  </tr>
	  <tr>
	    <td><code>--mapq</code></td>
	    <td>maximum allowed mapq score
                           calculated as the sum of qualities at base
                           mismatches</td>
	    <td>int</td>
	    <td>100</td>
	  </tr>
	  <tr>
	    <td><code>-u</code></td>
	    <td>only report reads different by an integer number of copy numbers from the reference allele</td>
	    <td>bool</td>
	    <td>-</td>
	  </tr>
	  <tr>
	    <td><code>-m<br>--mismatch</code></td>
	    <td>allowed edit distance in either flanking region</td>
	    <td>int</td>
	    <td>1</td>
	  </tr>
	  <tr>
	    <td><code>-g</code></td>
	    <td>Maximum number of gap opens allowed in each flanking region</td>
	    <td>int</td>
	    <td>1</td>
	  </tr>
	  <tr>
	    <td><code>-e</code></td>
	    <td>Maximum number of gap extends allowed in each flanking region</td>
	    <td>int</td>
	    <td>1</td>
	  </tr>
	  <tr>
	    <td><code>-r</code></td>
	    <td>fraction of missing alignments given 2% uniform base error rate. Ignored if -m is set</td>
	    <td>float</td>
	    <td>-1</td>
	  </tr>
	  <tr>
	    <td><code>--max-hits-quit-aln</code></td>
	    <td>Stop alignment search after this many hits found. Use -1 for no limit.</td>
	    <td>int</td>
	    <td>1000</td>
	  </tr>
	  <tr>
	    <td><code>--min-flank-allow-mismatch</code></td>
	    <td>don't allow mismatches if aligning flanking regions shorter than this</td>
	    <td>int</td>
	    <td>30</td>
	  </tr>
	</table>

<a name="allelotype"></a>
<h3>allelotype</h3>
<table class="box-table-a">
  <tr><th>Parameter</th><th>Description</th><th>Type</th><th>Default</th></tr>
  <tr>
    <td><code>--command</code></td>
    <td>(REQUIRED). One of:
      <ul>
	<li>train: train the noise model based on the provided BAM
	  alignment. This only works on male samples with a large number of
	  reads aligned to sex chromosome STRs. It is currently recommended
	  that you use the provided stutter model file. 
	  Training outputs the files:
	  <ul>
	    <li>$noisemodelprefix.stepmodel</li>
	    <li>$noisemodelprefix.stuttermodel</li>
	    <li>$noisemodelprefix.stutterproblem</li>
	  </ul>
	</li>
	<li>classify:
	  generate allelotypes from the provided alignment file and using the
	  provided noise model.
	  Classifying outputs the files:
	  <ul>
	    <li>$output_prefix.vcf</li>
	    <li>$output_prefix.genotypes.tab</li>
	  </ul>
	</li>
      </ul>
    </td>
    <td>string</td>
    <td>-</td>
  </tr>
  <tr>
    <td><code>--bam</code></td>
    <td>(REQUIRED) comma-separated list of BAM files to analyze. Each should have a unique read group and be sorted and indexed</td>
    <td>string</td>
    <td>-</td>
  </tr>
  <tr>
    <td><code>--out</code></td>
    <td>(REQUIRED) Prefix to name output files</td>
    <td>string</td>
    <td>-</td>
  </tr>
  <tr>
    <td><code>--strinfo</code></td>
    <td>(REQUIRED) File containing statistics for each STR. Available in the resource bundle download.</td>
    <td>string</td>
    <td>-</td>
  </tr>
  <tr>
    <td><code>--noise_model</code></td>
    <td>(REQUIRED) Prefix of files to write (--command train) or read (--command classify) noise model parameters to.</td>
    <td>string</td>
    <td>-</td>
  </tr>
  <tr>
    <td><code>--index-prefix</code></td>
    <td>(REQUIRED) prefix for lobSTR's bwa reference (must run lobstr_index.py to create index of download from lobSTR website). If the index is in PATH_TO_INDEX, this argument is PATH_TO_INDEX/lobSTR_.</td>
    <td>string</td>
    <td>-</td>
  </tr>
  <tr>
    <td><code>--no-rmdup</code></td>
    <td>Don't remove PCR duplicates before allelotyping.</td>
    <td>bool</td>
    <td>-</td>
  </tr>
    <tr>
    <td><code>--realign</code></td>
    <td>Redo local realignment. Useful if using alignments generated by other tools.</td>
    <td>bool</td>
    <td>-</td>
  </tr>
  <tr>
    <td><code>--min-het-freq</code></td>
    <td>minimum frequency to make a heterozygous call</td>
    <td>float</td>
    <td>0</td>
  </tr>
  <tr>
    <td><code>--haploid</code></td>
    <td>comma-separated list of chromosomes to force homozygous calls.</td>
    <td>string</td>
    <td>-</td>
  </tr>
  <tr>
    <td><code>-h<br>--help</code></td>
    <td>display help screen</td>
    <td>bool</td>
    <td>-</td>
  </tr>
  <tr>
    <td><code>-v<br>--verbose</code></td>
    <td>print out useful progress messages</td>
    <td>bool</td>
    <td>-</td>
  </tr>
  <tr>
    <td><code>--quiet</code></td>
    <td>don't print anything to stderr or stdout</td>
    <td>bool</td>
    <td>-</td>
  </tr>
  <tr>
    <td><code>--version</code></td>
    <td>print out lobSTR program version</td>
    <td>bool</td>
    <td>-</td>
  </tr>
  <tr>
    <td><code>--include-gl</code></td>
    <td>Include GL field in the VCF</td>
    <td>bool</td>
    <td>-</td>
  </tr>
  <tr>
    <td><code>--min-border</code></td>
    <td>Filter reads that do not extend past both ends of the STR region by this many bp.</td>
    <td>int</td>
    <td>5</td>
  </tr>
  <tr>
    <td><code>--mapq</code></td>
    <td>maximum allowed mapq score
      calculated as the sum of qualities at base
      mismatches</td>
    <td>int</td>
    <td>100</td>
  </tr>
  <tr>
    <td><code>--max-matedist</code></td>
    <td>Filter reads with a mate distance larger than this many bp.</td>
    <td>int</td>
    <td>100000</td>
  </tr>
  <tr>
    <td><code>--max-diff-ref</code></td>
    <td>only report reads differing by at most this number of bp from the reference allele</td>
    <td>int</td>
    <td>50</td>
  </tr>
  <tr>
    <td><code>--min-bp-before-indel</code></td>
    <td>Filter reads with an indel occurreing less than this many bp from eithe rend of the read</td>
    <td>int</td>
    <td>7</td>
  </tr>
  <tr>
    <td><code>--min-read-end-match</code></td>
    <td>Filter reads whose alignments don't exactly match the reference for at least this many bp at both ends</td>
    <td>int</td>
    <td>5</td>
  </tr>
  <tr>
    <td><code>--maximal-end-match</code></td>
    <td>Filter reads whose prefix/suffix matches to reference are less than or equal to those obtained when shifting the read ends by distances within this many bp.</td>
    <td>int</td>
    <td>15</td>
  </tr>
  <tr>
    <td><code>--chrom</code></td>
    <td>only processs loci on this chromosome.</td>
    <td>string</td>
    <td>-</td>
  </tr>
  <tr>
    <td><code>--unit</code></td>
    <td>only report reads different by an integer number of copy numbers from the reference allele</td>
    <td>bool</td>
    <td>-</td>
  </tr>
  <tr>
  	<td><code>--max-repeats-in-ends</code></td>
  	<td>Filter reads with more than this number of occurrences of the repeat motif in the 4*period bp on either end of the read. -1 means no filter is applied.</td>
  	<td>int</td>
  	<td>-1</td>
  </tr>
  <tr>
  	<td><code>--filter-mapq0</code></td>
  	<td>Filter reads with map quality 0. Only use for alignments not generated by lobSTR.</td>
  	<td>bool</td>
  	<td>-</td>
  </tr>
  <tr>
  	<td><code>--filter-clipped</code></td>
  	<td>Filter reads with hard or soft clipped bases at the ends.</td>
  	<td>bool</td>
  	<td>-</td>
  </tr>
  <tr>
  	<td><code>--filter-reads-with-n</code></td>
  	<td>Filter reads that have one or more N bases</td>
  	<td>bool</td>
  	<td>-</td>
  </tr>
  <tr>
    <td><code>--chunksize</code></td>
    <td>Number of loci to read into memory at a time</td>
    <td>int</td>
    <td>1000</td>
  </tr>
  <tr>
  	<td><code>--output-bams</code></td>
  	<td>Output BAM files:
  		<ul>
  			<li>out.reads.bam: contains all reads used for analysis (before collapsing duplicates)</li>
  			<li>out.filtered.bam: contains all reads removed by filters.</il>
  		</ul>
  		where out is the argument to <code>--out</code>.
  	</td>
  	<td>bool</td>
  	<td>-</td>
  </tr>
  <tr>
    <td><code>--noweb [deprecated]</code></td>
    <td>Do not report any user information to Amazon S3. [This feature has been discontinued]</td>
    <td>bool</td>
    <td>-</td>
  </tr>
</table>

<a name="index"></a>
<h3>lobstr_index.py</h3>
<table class="box-table-a">
  <tr><th>Parameter</th><th>Description</th><th>Type</th><th>Default</th></tr>
  <tr>
    <td><code>--str</code></td>
    <td>(REQUIRED) A bed file containing a list of STR loci
	     output from Tandem Repeat Finder. This file does not
	     necessarily need to be created by TRF. The necessary
	     columns are described above</td>
    <td>string</td>
    <td>-</td>
  </tr>
  <tr>
    <td><code>--ref</code></td>
    <td>(REQUIRED) A fasta file containing the reference genome</td>
    <td>string</td>
    <td>-</td>
  </tr>
  <tr>
    <td><code>--out_dir</code></td>
    <td>(REQUIRED) the directory to store the index created</td>
    <td>string</td>
    <td>-</td>
  </tr>
  <tr>
    <td><code>-h<br>--help</code></td>
    <td>Display a helpful usage message</td>
    <td>bool</td>
    <td>-</td>
  </tr>
  <tr>
    <td><code>-v<br>--verbose</code></td>
    <td>Display helpful progress messages</td>
    <td>bool</td>
    <td>-</td>
  </tr>
  <tr>
    <td><code>--extend</code></td>
    <td>The number of base pairs to include
      flanking either region of the STR. This should be at
      least as long as the read length you intend to
      align. IMPORTANT: if you change the extend parameter
      here, you MUST remember to set the --extend paramter to
      the same value when you call lobSTR to ensure that the output
      coordinates are accurate</td>
    <td>int</td>
    <td>1000</td>
  </tr>
</table>
      </section>
    </div>
    {% include footer.html %}
  </body>
</html>
