<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.1//EN" "http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
<title>FAQ - lobSTR</title>

<meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
<meta http-equiv="refresh" content="600"> 

<link rel='stylesheet' href='http://cdn1.editmysite.com/editor/images/common/common-v2.css?22' type='text/css' />
<link rel='stylesheet' type='text/css' href='files/main_style.css?1332804341' title='wsite-theme-css' />
<style type='text/css'>
#wsite-content div.paragraph, #wsite-content p, #wsite-content .product-description, .blog-sidebar div.paragraph, .blog-sidebar p, .wsite-form-field label, .wsite-form-field label {}
#wsite-content h2, #wsite-content .product-title, .blog-sidebar h2{}
#wsite-title{font-size:40px !important;}
</style>
<script type='text/javascript'><!--
var STATIC_BASE = 'http://cdn1.editmysite.com/';
var STYLE_PREFIX = 'wsite';
//-->
</script>
<script type='text/javascript' src='http://cdn1.editmysite.com/libraries/prototype/1.7-custom/prototype.min.js'></script>
<script type='text/javascript' src='http://cdn1.editmysite.com/libraries/scriptaculous/1.9.0-custom/effects.min.js'></script>
<script type='text/javascript' src='http://cdn1.editmysite.com/editor/images/common/utilities.js?17'></script>
<script type='text/javascript' src='http://cdn1.editmysite.com/editor/images/common/lightbox202.js?110'></script>
<script type='text/javascript' src='http://cdn1.editmysite.com/editor/libraries/flyout_menus.js?13'></script>
<script type='text/javascript'><!--
var IS_ARCHIVE=1;
function initFlyouts(){initPublishedFlyoutMenus([{"id":"155159388479881257","title":"Home","url":"index.html"},{"id":"482706786380939227","title":"Download","url":"download.html"},{"id":"911862752701740992","title":"Install","url":"install.html"},{"id":"705726613435707897","title":"Usage","url":"usage.html"},{"id":"495479748563597039","title":"File formats","url":"file-formats.html"},{"id":"262602591113460083","title":"News","url":"news.html"},{"id":"359371236662353445","title":"FAQ","url":"faq.html"},{"id":"854238589279126447","title":"Contact Us","url":"contact-us.html"}],'359371236662353445',"<li class='wsite-nav-more'><a href='#'>more...<\/a><\/li>",'active',false)}
if (Prototype.Browser.IE) window.onload=initFlyouts; else document.observe('dom:loaded', initFlyouts);
//-->
</script>
</head>
<script type="text/javascript">

  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', 'UA-30530333-1']);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();

</script>
<body class='wsite-theme-light wsite-page-faq'>
<div id="header-wrap"><div id="header-content">

		<h1 id="logo"><span class='wsite-logo'><a href='index.html'><img src='uploads/1/1/4/4/11449783/1332803980.png' style='margin-top:-2px;margin-left:-17px;max-height:67px;' /></a></span></h1>

		<ul><li id='pg155159388479881257'><a href='index.html'>Home</a></li><li id='pg482706786380939227'><a href='download.html'>Download</a></li><li id='pg911862752701740992'><a href='install.html'>Install</a></li><li id='pg705726613435707897'><a href='usage.html'>Usage</a></li><li id='pg495479748563597039'><a href='file-formats.html'>File formats</a></li><li id='pg262602591113460083'><a href='news.html'>News</a></li><li id='active'><a href='faq.html'>FAQ</a></li><li id='pg854238589279126447'><a href='contact-us.html'>Contact Us</a></li></ul>

	</div></div>

	<!-- content-wrap starts here -->
	<div id="content-wrap"><div id="content">

	<div id="main">

	<div id='wsite-content' class='wsite-not-footer'>
<div class='wsite-not-footer'>
<h2  style=" text-align: left; ">FAQ</h2>
<div  class="paragraph editable-text" style=" text-align: left; display: block; ">
<strong>How do I cite lobSTR?</strong><br/>
Please cite: <emph>Gymrek M, Golan D, Rosset S, & Erlich Y. lobSTR: A short tandem repeat profiler for personal genomes. Genome Research. 2012 April 22.</emph><br /><br />

<strong>Can lobSTR detect STRs of periods below 2 or above 6?</strong><br/>
The definition of what constitutes an STR is not clear cut. Currently lobSTR is limited to STR motifs of periods 2 through 6. We plan to include support for periods up to 8 in later versions. Much larger than that gets away from what is generally included in the "STR" category and is beyond the limits for detection by lobSTR's sensing step. Homopolymers have not been included because they are often quite noisy and difficult to type accurately. These may also be included in later versions.<br/><br/>

<strong>How are mismatches calculated?</strong><br />
The parameters that control alignment mismatches/gaps are:<ul style=""><li style="">-m&nbsp;: edit distance allowed</li><li style="">-r&nbsp;: the fraction of missing alignments given 2% uniform base error rate. Note if m is set, then r is ignored.</li><li style="">-g&nbsp;: the number of gap opens allowed</li><li style="">-e&nbsp;: the number of gap extends allowed</li></ul>Each flanking region is aligned separately, so each paramter above is applied to both the left and right flanking regions. That means, if you allow an edit distance of 2 (-m 2), then edit distance 2 is allowed both in the left and right flanking sequences.<br /><br />

<strong>How is the sam format generated?</strong><br />
After an STR alignment is found, lobSTR performs a local realignment using the Needleman Wunsch algorithm with an affine gap penalty score. This step allows lobSTR to generate accurate CIGAR scores for the entire read, including the STR region, that can be used downstream for viewing in samtools tview and for calling SNPs in STR regions.<br /><br />

<strong>How do you obtain the genomic coordinates of markers?</strong><br />
We use the table of STR's generated by&nbsp;<a href="http://tandem.bu.edu/trf/trf.html" style="">Tandem Repeat Finder</a>&nbsp;available on the UCSC Genome Browser.<br /><br />

<strong>How should I filter my results to obtain only high quality genotypes?</strong><br />
lobSTR reports a score for each genotype result based in the likelihood ratios of each possible genotype, ranging from 0 to 1. However, this score is currently unreliable in many cases and we do not recommend filtering on it with the current version. Several recommended ways to filter are on (1) the percentage of reads agreeing with the call (column 9 divided by column 8) (2) by the number of alleles reported (many alleles reported at the same locus are an indication that it is hard to call and may be unreliable) (3) by the coverage level. We found 20x to be a reliable coverage level in Illumina data, but calls at coverages as low as 3x were shown to give quite accurate results in many cases.<br /><br />

<strong>What does the remove duplicates procedure do? How can I turn it off?&nbsp;</strong><br />
All reads starting at the same 5' location with the same length are collapsed into a single read. If the reads have multiple copy numbers of an STR, the copy number exhibited by the majority of the PCR duplicates is retained. If there is no majority vote, quality scores are used to break ties. To turn off removing duplicates, use the --no-rmdup option to&nbsp;allelotype.<br /><br />

<strong>How can I turn off trimming?</strong><br />
lobSTR trims reads based on quality scores to remove erroneous read
ends. The --bwaq acts the same as the -q parameter for
	  BWA. Additionally, by default, lobSTR trims all flanking regions to 25bp before aligning. To avoid this step, set --maxflank to a value near the read length of the input.<br /><br/>

<strong>What does the --noise_model option take?</strong><br />
The --noise_model option takes a file name to read or write the noise model parameters to. If you specify --command classify, the allelotype program reads in the noise model file given and uses those parameters. If you specify --command train, allelotype will write to the file given to noise_model with the new trained parameters. Note that training is only reliable when the sample is male and contains a large number of reads aligned to sex chromosomes, so in most cases we recommend just using the --command classify option with a provided noise model.<br /><br/>

<strong>lobSTR output appears in the aligned.tab file, but not the aligned.bam file</strong><br />
In cases of very small input files (on the order of a few million reads) or specifiying a very large value to -p, the BAM output file is not properly written to. This is a known bug and will be fixed in later versions. In the case of small files, -p 1 will result in the BAM file being properly written. lobSTR has only been tested with values of -p of up to 40.<br /><br >

<strong>The nucleotides column reported in the aligned.tab and aligned.bam output files does not match the nucleotides of the original read.</strong><br />
This is a results of a bug in reporting this column in the initial
release. This should not affect STR call accuracy. This has been fixed
in downloads lobstr_v1.0.6_2 or later.<br/><br/>

<strong>The lobSTR indexer did not produce any output.</strong><br />
In versions below 2.0.0, lobSTR requires chromosome names to not
contain "_". If lobSTR does not return any output from the indexing
step, it may be because this illegal character is present. This has
been changed in version 2.0.0 to allow "_" but disallow "$" in
chromosome names.<br/><br/>

<strong>What read lengths can lobSTR align?</strong><br />
Because most STRs in lobSTR's reference are ~40bp, and a read must
fully span an STR plus some flanking region in order to produce a
unique alignment, lobSTR works best with reads of 100bp or more. Some high
quality alignments can be produced with reads as short as 45bp, but alignment errors are much more frequent with these short reads. 36bp
reads cannot be reliably aligned by lobSTR.
<br/><br/>



</div>
</div>
</div>


	</div>

	<!-- content-wrap ends here -->
	</div></div>

	<!-- footer starts here -->
	<div id="footer-wrap"><div id="footer-content">

		<div class="col float-left space-sep">
			<p>
			<span id='weeblyFooter'>Create a <a target="_top" href="http://www.weebly.com/">free web site</a> with <a target="_top" href="http://www.weebly.com/" title="free web site">Weebly</a></span>
			</p>
		</div>
		<br class="clear" />

	</div></div>
	<!-- footer ends here -->


</body>
</html>
